### PR TITLE
[FIX] uom: allow ratio change

### PR DESCRIPTION
**Pull request by ajf-odoo**
Allow unit of measure ratio change when there are no product currently reserved
or in stock using this unit of measure.

opw-2455914

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/product.py [hunk 1]
```diff
@@ -833,12 +833,25 @@ def write(self, values):
...
-                    ('state', '!=', 'cancel'),
+                mapped_ids = changed.mapped('id')
+                ok = False
+                stock_moves = self.env['stock.move'].search_count([
+                    ('product_uom', 'in', mapped_ids),
+                    ('state', 'not in', ('cancel', 'done')),
                 ])
-
-                if stock_move_lines:
+                if not stock_moves:
```
Comment (amoyaux) [inline addons/stock/models/product.py:842]: I would rather use a structure like
```python3
error_msg = _( "You cannot change the ratio of this unit of mesure as some"
" products with this UoM have already been moved or are "
"currently reserved."
))
if self.env['stock.move'].search_count([
    ('product_uom', 'in', mapped_ids),
    ('state', 'not in', ('cancel', 'done')
)]):
    raise UserError(error_msg)
if self.env['stock.move.line'].search_count([
    ('product_uom_id', 'in', mapped_ids),
    ('state', 'not in', ('cancel', 'done')),
]):
    raise UserError(error_msg)
if self.env['stock.quant'].search_count([
    ('product_id.product_tmpl_id.uom_id', 'in', mapped_ids),
    ('quantity', '!=', 0),
]):
    raise UserError(error_msg)
```
I find it more readable

File: addons/stock/models/product.py [hunk 2]
```diff
@@ -833,12 +833,25 @@ def write(self, values):
                 lambda u: any(u[f].id != int(values[f]) if f in values else False
                               for f in {'category_id'}))
             if changed:
-                stock_move_lines = self.env['stock.move.line'].search_count([
-                    ('product_uom_id.category_id', 'in', changed.mapped('category_id.id')),
-                    ('state', '!=', 'cancel'),
+                mapped_ids = changed.mapped('id')
+                ok = False
+                stock_moves = self.env['stock.move'].search_count([
```
Comment (fmdl) [inline addons/stock/models/product.py:838]: use sudo() (if multi company, ...)

File: addons/stock/models/product.py [hunk 3]
```diff
@@ -833,12 +833,25 @@ def write(self, values):
                 lambda u: any(u[f].id != int(values[f]) if f in values else False
                               for f in {'category_id'}))
             if changed:
-                stock_move_lines = self.env['stock.move.line'].search_count([
-                    ('product_uom_id.category_id', 'in', changed.mapped('category_id.id')),
-                    ('state', '!=', 'cancel'),
+                mapped_ids = changed.mapped('id')
```
Comment (fmdl) [inline addons/stock/models/product.py:836]: mapped_ids = changed.ids

**Comment by amoyaux** [comment]
robodoo r+

File: addons/stock/models/product.py (hunk 1 above)
Comment (mart-e) [reply addons/stock/models/product.py]: :point_up_2:

**Pull request closed without merging**